File: android/src/main/java/ch/nevis/mobile/authentication/sdk/react/model/message/incoming/RegistrationMessage.kt
/*
 * Copyright © 2023 Nevis Security AG. All rights reserved.
 */

package ch.nevis.mobile.authentication.sdk.react.model.message.incoming

import ch.nevis.mobile.authentication.sdk.react.model.serializer.AuthorizationProviderSerializer
import ch.nevis.mobile.authentication.sdk.react.model.serializer.DeviceInformationSerializer
import ch.nevis.mobile.authentication.sdk.react.model.serializer.PinPolicySerializer
import ch.nevis.mobile.authentication.sdk.react.model.serializer.RequestHeadersSerializer
import ch.nevis.mobile.sdk.api.authorization.AuthorizationProvider
import ch.nevis.mobile.sdk.api.localdata.DeviceInformation
import ch.nevis.mobile.sdk.api.operation.RequestHeaders
import ch.nevis.mobile.sdk.api.operation.pin.PinPolicy
import kotlinx.serialization.Serializable

@Serializable
data class RegistrationMessage(
    override val operationId: String,
    override val accountSelectorProvided: Boolean,
    override val authenticatorSelectorProvided: Boolean,
    override val pinEnrollerProvided: Boolean,
    override val pinUserVerifierProvided: Boolean,
    override val biometricUserVerifierProvided: Boolean,
    override val devicePasscodeUserVerifierProvided: Boolean,
    override val fingerprintUserVerifierProvided: Boolean,
    override val onSuccessProvided: Boolean,
    override val onErrorProvided: Boolean,
    @Serializable(with = RequestHeadersSerializer::class)
    override val requestHeaders: RequestHeaders? = null,
    val username: String? = null,
    val serverUrl: String? = null,
    @Serializable(with = DeviceInformationSerializer::class)
    val deviceInformation: DeviceInformation? = null,
    @Serializable(with = AuthorizationProviderSerializer::class)
    var authorizationProvider: AuthorizationProvider? = null,
    @Serializable(with = PinPolicySerializer::class)
    val pinPolicy: PinPolicy? = null,
    val allowClass2Sensors: Boolean? = null,
    val allowDevicePasscodeAsFallback: Boolean? = null,
    val invalidateOnNewOsBiometrics: Boolean? = null
) : UserInteractionOperationMessage()

File: ios/Model/Message/Incoming/RegistrationMessage.swift
//
// Copyright © 2023 Nevis Security AG. All rights reserved.
//

import NevisMobileAuthentication

class RegistrationMessage: UserInteractionOperationMessage {
	// MARK: Properties

	var username: String?
	var serverUrl: String?
	var pinPolicy: PinPolicy?
	var deviceInformation: DeviceInformation?
	var authorizationProvider: TypedAuthorizationProvider?
	var allowDevicePasscodeAsFallback: Bool?
	var invalidateOnNewOsBiometrics: Bool?

	// MARK: Decodable

	enum CodingKeys: String, CodingKey {
		case username
		case serverUrl
		case pinPolicy
		case deviceInformation
		case authorizationProvider
		case allowDevicePasscodeAsFallback
		case invalidateOnNewOsBiometrics
	}

	required init(from decoder: Decoder) throws {
		let container = try decoder.container(keyedBy: CodingKeys.self)
		self.username = try container.decodeIfPresent(String.self, forKey: .username)
		self.serverUrl = try container.decodeIfPresent(String.self, forKey: .serverUrl)
		self.pinPolicy = try container.decodeIfPresent(PinPolicy.self, forKey: .pinPolicy)
		self.deviceInformation = try container.decodeIfPresent(DeviceInformation.self, forKey: .deviceInformation)
		self.authorizationProvider = try container.decodeIfPresent(TypedAuthorizationProvider.self, forKey: .authorizationProvider)
		self.allowDevicePasscodeAsFallback = try container.decodeIfPresent(Bool.self, forKey: .allowDevicePasscodeAsFallback)
		self.invalidateOnNewOsBiometrics = try container.decodeIfPresent(Bool.self, forKey: .invalidateOnNewOsBiometrics)
		try super.init(from: decoder)
	}
}
